src/plugin.js

```diff
@@ -426,57 +35,82 @@ const findRoot = (className) => {
-  const player = this;
-
-  if (!player.playlist) {
-    throw new Error('videojs-playlist plugin is required by the videojs-playlist-ui plugin');
-  }
-
-  options = merge(defaults, options);
+class PlaylistUI extends Plugin {
 
-  // If the player is already using this plugin, remove the pre-existing
-  // PlaylistMenu, but retain the element and its location in the DOM because
-  // it will be re-used.
-  if (player.playlistMenu) {
-    const el = player.playlistMenu.el();
+  constructor(player, options) {
+    super(player, options);
 
-    // Catch cases where the menu may have been disposed elsewhere or the
-    // element removed from the DOM.
-    if (el) {
-      const parentNode = el.parentNode;
-      const nextSibling = el.nextSibling;
+    if (!player.usingPlugin('playlist')) {
+      player.log.error('videojs-playlist plugin is required by the videojs-playlist-ui plugin');
+      return;
+    }
 
-      // Disposing the menu will remove `el` from the DOM, but we need to
-      // empty it ourselves to be sure.
-      player.playlistMenu.dispose();
-      dom.emptyEl(el);
+    options = this.options_ = videojs.obj.merge(defaults, options);
 
-      // Put the element back in its place.
-      if (nextSibling) {
-        parentNode.insertBefore(el, nextSibling);
-      } else {
-        parentNode.appendChild(el);
-      }
-
-      options.el = el;
+    if (!videojs.dom.isEl(options.el)) {
+      options.el = this.findRoot_(options.className);
     }
+
+    // Expose the playlist menu component on the player as well as the plugin
+    // This is a bit of an anti-pattern, but it's been that way forever and
+    // there are likely to be integrations relying on it.
+    this.playlistMenu = player.playlistMenu = new PlaylistMenu(player, options);
   }
 
-  if (!dom.isEl(options.el)) {
-    options.el = findRoot(options.className);
+  /**
+   * Dispose the plugin.
+   */
+  dispose() {
+    super.dispose();
+    this.playlistMenu.dispose();
   }
 
-  player.playlistMenu = new PlaylistMenu(player, options);
-};
+  /**
+   * Returns a boolean indicating whether an element has child elements.
+   *
+   * Note that this is distinct from whether it has child _nodes_.
+   *
+   * @param  {HTMLElement} el
+   *         A DOM element.
+   *
+   * @return {boolean}
+   *         Whether the element has child elements.
+   */
+  hasChildEls_(el) {
+    for (let i = 0; i < el.childNodes.length; i++) {
+      if (videojs.dom.isEl(el.childNodes[i])) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Finds the first empty root element.
+   *
+   * @param  {string} className
+   *         An HTML class name to search for.
+   *
+   * @return {HTMLElement}
+   *         A DOM element to use as the root for a playlist.
+   */
+  findRoot_(className) {
+    const all = document.querySelectorAll('.' + className);
+    let el;
+
+    for (let i = 0; i < all.length; i++) {
+      if (!this.hasChildEls_(all[i])) {
```

Comment:
nit: Could this be improved by starting at length - 1 and returning as soon as we find it? We would not need the `el` variable in this case either

Reply:
Not exactly since we need to find the _first_ matching... but we could simply return from inside the `for` loop!